### DISCUSSION TITLE
Generator models with Simulink/dll

### DISCUSSION THREAD

**Discussion by FerminVelazquez** [OpenFAST/openfast#1202] [Q&A]
Hi everyone,

I am trying to simulate grid faults (shortcircuits, voltage dips, etc) and ultimately their effect in mechanical loads (to the tower, for example). To do so, I would like to include a dynamical model of a generator, implemented in Simulink (in a DFIG or Full Converter configuration), that models the transient behaviour of the whole system (from grid to electrical torque). 

I found that it can be included in the ServoDyn module, within the UserGen subroutine. The thing is, could it be implemented as a .dll? 

I don’t know if it is possible to call the dll in the UserGen subroutine (which is defined in the UserVSCont_KP.f90 file), or if it should be done in other way. 

I assume that in this way, the VSContrl should be set to zero, as the control is implemented in the model, right?

![cinfiguracion DFIG](https://user-images.githubusercontent.com/110229232/181793101-220a59a1-6811-4dee-b544-d12eba458b62.png)

The model simulates all the parts shown in the picture. That is, the generator itself, the converters, the grid and the controllers of the converters. The output is the electric torque. Currently, I have as inputs mechanical torque (given by the turbine), speed reference (fed to the rotor-side converter controller), electric torque reference (which is fed to the converter control blocks within the Simulink model) and grid voltage.

Thank you in advance.

Kind regards,

**Comment by jjonkman** [comment]
Dear @FerminVelazquez,

I'm not sure I fully understand your question, but please note that `VSContrl` and `GenModel` are, in the end, just different ways of specifying the electrical generator torque.  If you already have a generator model implemented in Simulink, you can interface that to OpenFAST by selecting `VSContrl` = 4 (even if the Simulink model represents a generator instead of a torque controller).

If that doesn't help, please clarify your question.

FYI: The following NREL report may be of interest to you: https://www.nrel.gov/docs/fy14osti/59195.pdf.

Best regards,
